File: Model/latifiniti.cpp
#include "latifiniti.h"
#include <QVector>


LatiFiniti::LatiFiniti(const QVector<Punto*>& v) : vertici(v) {
}

QVector<double> LatiFiniti::getLati() const {
    QVector<double> t;
    for(QVector<Punto*>::const_iterator it=vertici.begin(); it != vertici.end()-1; ++it) {
        t.push_back(Punto::distanzaDuePunti(**it,**(it+1)));
    }
    t.push_back(Punto::distanzaDuePunti(*(vertici.first()),*(vertici.last())));
    return t;
}
double LatiFiniti::getPerimetro() const {
    double perim = 0;
    QVector<double> lati = getLati();
    for(QVector<double>::const_iterator it= lati.begin();it!=lati.end();++it) {
        perim += *it;
    }
    return perim;
}

unsigned int LatiFiniti::contaVertici() const {
    unsigned int nVertici = 0;
    for(QVector<Punto*>::const_iterator cit=vertici.begin(); cit != vertici.end(); ++cit) {
        nVertici++;
    }
    return nVertici;
}

Punto LatiFiniti::getBaricentro() const {
    unsigned int nVertici = contaVertici();
    double numerX = 0;
    double numerY = 0;
    for(unsigned int i=0; i<nVertici; ++i) {
        numerX += vertici[i]->getX();
        numerY += vertici[i]->getY();
    }
    return Punto(numerX/nVertici,numerY/nVertici);

}

double LatiFiniti::getLato(unsigned int l) const {
    return getLati()[l];
}

Punto LatiFiniti::getVertice(unsigned int i) const {
    return *vertici[i];
}

void LatiFiniti::setVertice(unsigned int i, const Punto& p) {
    const_cast<Punto*>(vertici[i])->setX(p.getX());
    const_cast<Punto*>(vertici[i])->setY(p.getY());
}

void LatiFiniti::pushVertice(const Punto & p) {
    vertici.push_back(new Punto(p));
}

File: SuperQtKalkJAVA/src/LatiFiniti.java
import java.util.ArrayList;

abstract class LatiFiniti implements Poligono {
    private ArrayList<Punto> vertici;

    public LatiFiniti(ArrayList<Punto> l) {
        vertici = l;
    }

    public ArrayList<Double> getLati() {
        ArrayList<Double> t = new ArrayList<Double>();
        for(int i = 0; i <vertici.size(); ++i) {
            t.add(Punto.distanzaDuePunti(vertici.get(i),vertici.get(i+1)));
        }
        t.add(Punto.distanzaDuePunti(vertici.get(0),vertici.get(vertici.size()-1)));
        return t;
    }

    public double getPerimetro() {
        double perim = 0;
        ArrayList<Double> lati = getLati();
        for(int i = 0; i < vertici.size(); ++i) {
            perim += lati.get(i);
        }
        return perim;
    }

    public int contaVertici() {
        int nVertici = 0;
        for(int i = 0; i < vertici.size(); ++i) {
            nVertici++;
        }
        return nVertici;
    }

    public Punto getBaricentro() {
        int nVertici = contaVertici();
        double numerX = 0;
        double numerY = 0;
        for(int i = 0; i < nVertici; ++i) {
            numerX += vertici.get(i).getX();
            numerY += vertici.get(i).getY();
        }
        return new Punto(numerX / nVertici, numerY / nVertici);
    }

    public double getLato(int i) {
        return getLati().get(i);
    }

    public Punto getVertice(int i) {
        return vertici.get(i);
    }

    public void setVertice(int i, Punto p) {
        vertici.get(i).setX(p.getX());
        vertici.get(i).setY(p.getY());
    }

    protected void pushVertice(Punto p) { //controllare se serva una new
        vertici.add(p);
    }
}
